Add export function to API

jd565 opened the pull request:
Add an export function that writes transaction data to a file and
returns a URI to access that file. The output format is the same as when
exporting from the chucker UI, but can be triggered from code.

## :camera: Screenshots
All internal changes - nothing different in UI

## :page_facing_up: Context
Adds functionality to export transactions to a file on the API. [Related issue](https://github.com/ChuckerTeam/chucker/issues/783)

## :pencil: Changes
Updated the Chucker API object to add in the new functionality. I've not made this a suspending function to maintain interop with java, so have just left a comment that this needs to be called on a background thread.

I added a button to the sample app (I had to comment out the strict mode checking otherwise the sample app closed immediately) which calls the new method - I'm not sure if this should be left in the sample long term so happy to remove it if it doesn't make sense.
<!-- Which code did you change? How? -->
<!-- If your changes affect users somehow, be it a new feature, a bugfix or an API change please update the `Unreleased` section in the CHANGELOG.md file. -->

## :paperclip: Related PR
<!-- PR that blocks this one, or the ones blocked by this PR -->

## :no_entry_sign: Breaking
<!-- Is there something breaking the API? Any class or method signature changed? -->
No

## :hammer_and_wrench: How to test
<!-- Is there a special case to test your changes? -->
Added a button to the sample app to export to a file then pulled the file and verified the contents.

## :stopwatch: Next steps
<!-- Do we have to plan something else after the merge? -->

cortinico commented on CHANGELOG.md:
```diff
@@ -11,6 +11,7 @@ Please add your entries according to this format.
 * Brotli compression support
 * Added ability to export list of transactions as .har file.
 * Added ability to save single transaction as .har file.
+* Added ability to export transactions to a file triggered in code.
```
```suggestion
* Added ability to export transactions to a file programmatically.
```

cortinico commented on library-no-op/src/main/kotlin/com/chuckerteam/chucker/api/Chucker.kt:
```diff
@@ -19,4 +22,11 @@ public object Chucker {
         // Empty method for the library-no-op artifact
     }
+
+    @JvmStatic
+    public fun writeTransactions(
+        context: Context,
+        maxTransactions: Long?,
+        startTimestamp: Long?,
+    ): Uri? = null
```
I think this should be moved to `ChuckerCollector`

jd565 replied:
I have updated this to be inside ChuckerCollector, but it seems slightly odd to me - (as a consumer) why do I have to instantiate a class that is about handling sending/receiving http messages in order to dump the saved transactions to a file?

cortinico replied:
The documentation says:

```
 * The collector responsible of collecting data from a [ChuckerInterceptor] and
 * storing it/displaying push notification. 
```

While the `Chucker` class you used originally is just the entry point of our library. As you want to do an advanced operation, I think it's reasonable to ask the user to instanciate and extra object. This will not pollute the `Chucker` interface making it a god object in the long run.

cortinico commented on library/src/main/kotlin/com/chuckerteam/chucker/api/Chucker.kt:
```diff
@@ -75,6 +81,43 @@ public object Chucker {
+        maxTransactions: Long?,
+        startTimestamp: Long?,
+    ): Uri? {
+        RepositoryProvider.initialize(context)
```
If you move this inside the Collector, this won't be needed.

cortinico commented on library/src/main/kotlin/com/chuckerteam/chucker/api/Chucker.kt:
```diff
@@ -75,6 +81,43 @@ public object Chucker {
+        RepositoryProvider.initialize(context)
+        val transactions =
+            RepositoryProvider.transaction().getTransactions(maxTransactions, startTimestamp)
+        if (transactions.isEmpty()) {
+            Toast
+                .makeText(context, R.string.chucker_export_empty_text, Toast.LENGTH_SHORT)
+                .show()
+            return null
+        }
```
This method should not show UI operations. It's up to the consumer of this method to do so if they wish (say if the URI returned is null).

cortinico commented on library/src/main/kotlin/com/chuckerteam/chucker/api/Chucker.kt:
```diff
@@ -75,6 +81,43 @@ public object Chucker {
+        }
+
+        val sharableTransactions = TransactionListDetailsSharable(transactions, encodeUrls = false)
+        return runBlocking {
```
This is not great as it might break the UI thread. Potentially returning a suspending function here would be better

cortinico commented on sample/src/main/kotlin/com/chuckerteam/chucker/sample/MainActivity.kt:
```diff
@@ -38,6 +38,13 @@ class MainActivity : AppCompatActivity() {
 
+            exportToFile?.visibility = if (Chucker.isOp) View.VISIBLE else View.GONE
+            exportToFile?.setOnClickListener {
+                Thread {
+                    Chucker.writeTransactions(this@MainActivity, 1, null)
+                }.start()
```
Please don't start a thread explicity here. We use coroutines to do any async work.

jd565 replied:
Have updated to use coroutines, but have had to add coroutines and activity-ktx dependencies to the sample gradle project as they weren't there before

cortinico replied:
Yeah I realized this is a side effect 🤔 I'll be fine with it. I'm unsure what's the other maintainers take on this (specifically @MiSikora )

cortinico commented on library/src/main/kotlin/com/chuckerteam/chucker/internal/data/repository/HttpTransactionDatabaseRepository.kt:
```diff
@@ -42,4 +42,13 @@ internal class HttpTransactionDatabaseRepository(private val database: ChuckerDa
 
     override suspend fun getAllTransactions(): List<HttpTransaction> = transactionDao.getAll()
+
+    override fun getTransactions(
```
```suggestion
    override fun getTransactionsInTimeRange(
```

or somethign more meaningful

cortinico commented on library/src/main/kotlin/com/chuckerteam/chucker/internal/data/repository/HttpTransactionRepository.kt:
```diff
@@ -26,4 +26,6 @@ internal interface HttpTransactionRepository {
 
     suspend fun getAllTransactions(): List<HttpTransaction>
+
+    fun getTransactions(maxTransactions: Long?, minTimestamp: Long?): List<HttpTransaction>
```
Same as above `getTransactionsInTimeRange`

cortinico commented on library/src/main/kotlin/com/chuckerteam/chucker/internal/data/room/HttpTransactionDao.kt:
```diff
@@ -44,4 +44,7 @@ internal interface HttpTransactionDao {
     suspend fun getAll(): List<HttpTransaction>
+
+    @Query("SELECT * FROM transactions WHERE requestDate >= :timestamp LIMIT :limit")
+    fun getTransactions(limit: Long, timestamp: Long): List<HttpTransaction>
```
Same as above: `getTransactionsInTimeRange`

cortinico commented on library/src/main/kotlin/com/chuckerteam/chucker/api/Chucker.kt:
```diff
@@ -75,6 +81,43 @@ public object Chucker {
+     * This must be called on a background thread.
+     *
+     * @param context Application context
+     * @param maxTransactions Maximum number of transactions to return in the file. Passing null
+     * means no limit on the number of transactions
```
Why would want to limit this? I don't fully understand the use case.

jd565 replied:
I'm not sure I have a specific use case but it seems like a good idea to have a limit call on an API like this.

Dumping the whole list of transactions from the database can take considerable time, so having ways to limit it and make this API a bit quicker would be good imo.

With the API like this users can specify either a time limit (e.g. calculate the last 1 hour) or can set a count limit (e.g. show the last 100 HTTP messages)

cortinico replied:
I would not add them, till we have a valid use case. We can still add them afterwards if needed. Having the method as suspending allows it to run on a separate thread, so having it as a long operation is not a major deal imho.

Between the two parameters, the timestamp seems more reasonable. The limit instead, will be useless if not paired with a proper pagination support (as you'll be essentially losing some transactions).

cortinico reviewed (commented):
Thanks for the PR. Left several comments that needs to be addressed before we can proceed

cortinico commented:
> I've not made this a suspending function to maintain interop with java, so have just left a comment that this needs to be called on a background thread.

Just noticed this. I would say that we can safely make this a suspending function at this stage.

Moreover, tests are missing. Could you add some of them?

jd565 commented:
> 

I think I have covered all your comments. Could you give some more information about what functions you are expecting to be tested?

The old `Sharable.shareAsFile` function has no tests on it, so I am not sure if you want that to be tested? The overall function aswell in ChuckerCollector I am not sure if you want a test on but I am not 100% sure how to achieve this without mocking everything and leaving a not very useful test.

jd565 commented:
@cortinico Thanks for looking over it. I think I have addressed all the comments above, but I'm not 100% sure about the testing side of it. Could you clarify what parts you think should have unit test coverage and I can look into it?

jd565 commented:
@cortinico Can you give any advice about the above?

vbuberen commented on library-no-op/src/main/kotlin/com/chuckerteam/chucker/api/Chucker.kt:
```diff
@@ -2,6 +2,9 @@ package com.chuckerteam.chucker.api
 import android.content.Intent
+import android.net.Uri
+import android.widget.Toast
+import com.chuckerteam.chucker.R
```
Do we need these new imports here?

vbuberen commented on library/src/main/kotlin/com/chuckerteam/chucker/api/Chucker.kt:
```diff
@@ -5,13 +5,19 @@ import android.content.Intent
+import android.widget.Toast
 import androidx.core.content.getSystemService
 import com.chuckerteam.chucker.R
+import com.chuckerteam.chucker.internal.data.repository.RepositoryProvider
```
I don't think we need these imports as well. Feels like some leftovers from experimenting, etc.

vbuberen reviewed (commented):
The approach looks fine to me.

Have left a few comments about strange new imports.

Also, I feel that we need to mention this feature in our Readme file as well.

BluestormDNA commented:
@jd565 @cortinico cound it be possible to add HAR file export too?

cortinico commented:
> @cortinico I also forgot about this :/ Your changes look fine and if you are happy to merge then I'm happy to.

Yup I'm merging this one 👍 Thank you again @jd565 

> @jd565 @cortinico cound it be possible to add HAR file export too?

@BluestormDNA please open a separate issue
